server/admin_management/api/app/services/user_service.py

```diff
@@ -19,32 +19,81 @@ def get_user_by_domain_and_name(self, user_type_code: str, user_name: str):
         return self.user_repo.get_users()
 
-    def find_or_create(self, user_type_code: str, user_name: str, requester: str):
+    def find_or_create(
+        self, user_type_code: str, user_name: str, user_guid: str, requester: str
+    ):
         LOGGER.debug(
             f"Request for finding or creating a user with user_type: {user_type_code}, "
             + f"user_name: {user_name}."
         )
 
-        fam_user = self.get_user_by_domain_and_name(user_type_code, user_name)
+        fam_user = self.get_user_by_domain_and_guid(user_type_code, user_guid)
         if not fam_user:
-            request_user = FamUserDto(
-                **{
-                    "user_type_code": user_type_code,
-                    "user_name": user_name,
-                    "create_user": requester,
-                }
+            fam_user_by_domain_and_name = self.get_user_by_domain_and_name(
+                user_type_code, user_name
             )
+            # if not found user by user_guid, but found user by domain and name
+            # in the case of historical FAM user that has no user_guid stored
+            # add their user_guid
+            if (
+                fam_user_by_domain_and_name
+                and not fam_user_by_domain_and_name.user_guid
+            ):
+                self.user_repo.update(
+                    fam_user_by_domain_and_name.user_id,
+                    {models.FamUser.user_guid: user_guid},
+                    requester,
+                )
+                LOGGER.debug(
+                    f"User {fam_user_by_domain_and_name.user_id} found, updated to store their user_guid."
+                )
+                return fam_user_by_domain_and_name
+            # if not found user by user_guid, and not found user by domain and username,
+            # or found user by domain and username, but user guid does not match (this is the edge case that could happen when username changed from IDP provider)
+            # create a new user
+            else:
+                request_user = FamUserDto(
+                    **{
+                        "user_type_code": user_type_code,
+                        "user_name": user_name,
+                        "user_guid": user_guid,
+                        "create_user": requester,
+                    }
+                )
 
-            fam_user = self.user_repo.create_user(request_user)
-            LOGGER.debug(f"User created: {fam_user.user_id}.")
-            return fam_user
+                fam_user_new = self.user_repo.create_user(request_user)
+                LOGGER.debug(f"User created: {fam_user_new.user_id}.")
+                return fam_user_new
 
         LOGGER.debug(f"User {fam_user.user_id} found.")
+        # update user_name if needs
+        fam_user = self.update_user_name(fam_user, user_name, requester)
         return fam_user
 
+    def update_user_name(
+        self,
+        user: models.FamUser,
+        user_name_from_request: str,
```

Comment:
Maybe "user_name_for_update" sounds better?

Reply:
so here we try to compare the username in our database vs. the username from the request, if not the same, then we update. I feel "user_name_from_request" makes more sense?

Reply:
I feel the variable should not be context aware because the method itself provide the logic to do the job and does not need to care who passes the value (from request, or from other method). The calling method needs to know what it needs to pass. Also this probably is a private method that's not meant for use from router/request (although the inputs most often are from the request).